Dumba Step04 - 포커딜러와 게임테스트

ghis22130 opened the pull request:
- shuffle을 직접 구현해 보았습니다. [참고](https://hyerios.tistory.com/18)

- Step03 피드백을 최대한 반영해보았습니다.

- 카드를 나눠주다가 카드가 다떨어지면 게임 종료 하는 부분이 구현하면서 난관을 많이 겪은 것 같습니다. 하나 씩 나눠주다 떨어지면 끝내는게 맞는지, 미리 계산해서 하는게 최선인지 어려웠습니다. 앞서 구현한 removeOne이 옵셔널 반환식이라 하나씩 흐름 따라가는게 어려웠습니다.

- Player는 값의 참조가 필요해서 Class 를 이용하여 구현해보았습니다.

- 구현에만 몰두 하다 보니 어느새 커다란 Commit 덩어리들이 되어있는 것을 깨달았습니다. 앞으로는 코드 작성시 계속 나누는 연습을 자연스러워 질때까지 의식하며 할것 같습니다.

ghis22130 committed:
Shuffle 직접 구현

ghis22130 committed:
Step03 피드백 반영

ghis22130 committed:
Dealer Struct 구현

ghis22130 committed:
Player Class 구현

ghis22130 committed:
PockerGame Struct 구현

ghis22130 committed:
CardDeck 추가 메소드 구현

isDeckEmty() -> Bool
fiveStudCards() -> [Card]?
sevenStudCards9) -> [Card]?

ghis22130 committed:
Step04 진행

ghis22130 committed:
카드가 부족하면 게임 종료 구현

카드 덱이 다 떨어질때까지 반복하여 그라운드를 진행한다.
카드 덱이 떨어지면 게임을 종료한다.

ghis22130 committed:
isDeckEmpty 메소드 삭제

ghis22130 committed:
Test Case 추가

ghis22130 committed:
Step04 README.md 작성

godrm commented:
> * 구현에만 몰두 하다 보니 어느새 커다란 Commit 덩어리들이 되어있는 것을 깨달았습니다. 앞으로는 코드 작성시 계속 나누는 연습을 자연스러워 질때까지 의식하며 할것 같습니다.

그래서 그런 방법을 쓰기도 합니다. 작업할 내용을 정리하고, 커밋 로그를 먼저 작성하고 그 일만 처리하고 커밋하는 거죠 ㅎㅎ

godrm commented on PokerGameApp/PokerGameApp/Card.swift:
```diff
@@ -33,7 +29,10 @@ enum Rank: Int, CustomStringConvertible, CaseIterable{
-    }
+        }
+    
+    case A = 1, Two, Three, Four, Five, Six, Seven, Eight, Nine, Ten, J, Q, K
```
case 보다도 더 description이 중요하다는 의미일까요? 위에서부터 읽기 때문에 윗쪽에 더 중요한 값이나 함수를 선언하는 게 좋습니다.

godrm commented on PokerGameApp/PokerGameApp/CardDeck.swift:
```diff
@@ -27,16 +27,49 @@ struct CardDeck {
+        return newCards
+    }
+    
+    mutating func fiveStudCards() -> [Card]?{
```
sevenStudCards() 와 fiveStudCards() 나눠진 것은 조금 어색합니다. 
예를 들어 둘 다 처리할 수 있는 `give(number ofCards:Int) -> [Card]` 는 어떨까요?

godrm commented on PokerGameApp/PokerGameApp/CardDeck.swift:
```diff
@@ -27,16 +27,49 @@ struct CardDeck {
+        for _ in 0...6 {
+            guard let newCard = self.removeOne() else {
+                print("카드 부족 게임 종료")
+                return nil
```
배열을 리턴할 때는 옵셔널nil 과 빈 배열의 차이를 구분해야 합니다. 이 함수는 어떤 차이가 있나요? 
CardDeck이 게임이 종료됐다는 로직까지 갖고 있어야 할까요? 카드덱은 카드만 관리하면 되는게 아닐까요?

godrm commented on PokerGameApp/PokerGameApp/Dealer.swift:
```diff
@@ -0,0 +1,49 @@
+
+struct Dealer : CustomStringConvertible{
+    
+    var description: String { return "\(cards)" }
```
description을 제공하는 것은 좋은 방법입니다만 문자열로 리턴하는 것은 좋은 방법이 아닙니다. 
이렇게 하면 앱 화면에 카드를 출력해야 하는데 이 정보는 아무런 도움이 안되니까요.

ghis22130 replied:
내부 동작 설명 처럼 구현해보고자해서 description을 이용했는데, 그러면
```swift
    func printPlayerDeck() {
        print(cards)
    }
```
이런식으로 접근하는것도 문제가 될까요?

godrm commented on PokerGameApp/PokerGameApp/Dealer.swift:
```diff
@@ -0,0 +1,49 @@
+    
+    var description: String { return "\(cards)" }
+    var cardDeck: CardDeck
+    var cards:[Card] = []
```
내부 속성을 다룰 때는 항상 private var를 우선적으로 고려하세요.

godrm commented on PokerGameApp/PokerGameApp/Dealer.swift:
```diff
@@ -0,0 +1,49 @@
+    
+    mutating func distributeCard(with player: Player, stud: Int) -> Bool{
+        switch stud {
+        case 5:
+            guard let newCards = cardDeck.fiveStudCards() else {return false}
+            player.receiveCards(with: newCards)
+        case 7:
+            guard let newCards = cardDeck.sevenStudCards() else {return false}
+            player.receiveCards(with: newCards)
+        default:
+            return false
+        }
+        return true
+    }
+    
+    mutating func takeDealerCards(stud: Int) -> Bool {
+        switch stud {
+        case 5:
+            guard let newCards = cardDeck.fiveStudCards() else {return false}
+            self.cards = newCards
+        case 7:
+            guard let newCards = cardDeck.sevenStudCards() else {return false}
+            self.cards = newCards
```
두 함수에서 모두 5, 7에 따라서 분기가 일어나는 데 이런 동작이 전형적인 추상 데이터 타입 방식입니다.
CS10 OOP 수업할 때 언급했던 자료를 찾아보세요. 
특정값에 따라서 분기하기 보다는 특정한 값/객체에 따라서 다형성으로 동작하도록 작성하는 게 더 좋습니다.

godrm commented on PokerGameApp/PokerGameApp/PockerGame.swift:
```diff
@@ -0,0 +1,51 @@
+struct PockerGame {
+    var players: [Player] = []
+    var dealer: Dealer
+    var cardStud: Int
```
여기 PockerGame에서 사용하는 players 타입 [Player]를 포함하는 Players 타입을 선언해보세요.
CardDeck이 [Card] 배열을 포함하는 것처럼 [Player] 배열을 다루는 Players 타입을 구현해보세요.
그러면 여기서 배열에 넣거나, 찾거나, 비교하거나 탐색하는 반복문들이 모두 Players로 옮겨질 겁니다.
그렇게 개선해보세요.

godrm commented:
PR 열어놓을 테니까 같은 브랜치에 수정사항을 PUSH하세요.
모두 수정하고 댓글을 남겨주세요.

ghis22130 committed:
description 위치조정

ghis22130 committed:
접근지정자 조정

ghis22130 committed:
Step04 피드백 적용

Players Class를 구현하여 player와 관련된 메소드 작성
추상데이터방식 수정
Dealer, Player 의 description을 제거하고 덱을 콘솔에서 보고싶을때 선언할 수 있는 함수 작성

ghis22130 commented:
- Players Class를 구현해보았습니다. 이 클래스는 플레이어들을 포함하고 있으며 플레이어들에게 딜러로부터 받은 덱을 나눠주는 기능을 하고 있습니다.

- struct type을 사용하다보니 객체간 교류가 쉽지 않아 class로 수정하였습니다.

- 추상 데이터 방식을 수정하였습니다. 포커 게임 자체가 5아니면 7로 한정되어있어 분기가 일어나도록 작성했었던 것 같습니다. 이는 객체 지향에서 지양해야할 방식임을 인지했습니다.  

- 게임 종료 부분 또한 적절한 곳에서 선언 해줘야 하는 것이 중요하다는 것을 알았습니다.

- 내부 속성에 접근지정자를 모두 지정해주었습니다. 

- 피드백을 수정해보고자 하나씩 건드려 commit해보는게 목표였는데 하나 건드리다보니 다른 부분도 건드려서 잘게 commit하기란.. 정말 어려운것 같습니다..ㅜ

godrm commented:
> * 피드백을 수정해보고자 하나씩 건드려 commit해보는게 목표였는데 하나 건드리다보니 다른 부분도 건드려서 잘게 commit하기란.. 정말 어려운것 같습니다..ㅜ

네. 그럴수 있죠. 그래도 커밋 하나에 한 가지 이슈에 대해서만 수정하려고 노력하시면 됩니다.

godrm commented on PokerGameApp/PokerGameApp/Card.swift:
```diff
@@ -8,12 +8,15 @@
-    
     case spades = "♠️", hearts = "❤️", diamonds = "♦️", clubs = "♣️"
+    
+    var description: String { return "\(self.rawValue)"}
```
스위프트는 한 줄 범위더라도 풀어서 쓰는 것을 권장합니다. 참고하세요.

godrm commented on PokerGameApp/PokerGameApp/Card.swift:
```diff
@@ -41,6 +41,11 @@ struct Card: CustomStringConvertible {
-    let rank: Rank
-    let suit: Suit
+    private let rank: Rank
+    private let suit: Suit
```
보통 상단이 더 중요한 정보를 배치하고 아래로 읽을 수록 부가적인 정보를 위치합니다.
Card 구조체에서 가장 중요한 것은 description일까요? rank, suit 일까요?

godrm commented on PokerGameApp/PokerGameApp/Dealer.swift:
```diff
@@ -9,41 +9,36 @@ import Foundation
-        default:
-            return false
-        }
+    func distributePlayers(with players: Players) -> Bool{
```
`func distributePlayers(with players: Players) -> Bool` 에서 Players 가 몇 번 반복되는지 읽어보세요.
함수 선언이나 호출할 때도 단어가 반복되지 않는 게 좋습니다.

godrm commented on PokerGameApp/PokerGameApp/Player.swift:
```diff
@@ -12,16 +12,14 @@ import Foundation
-    
+    func playerDeck() -> [Card] {
+        cards
+    }
```
아쉬운 부분은 Dealer와 Player가 공통점이 많은데 서로 전혀 코드가 공유하지 않는 점이네요. 
동작이 같으면 메소드를 일반화시켜서 구현할 수 있습니다. 다음 스텝을 하면서 이 부분을 꼭 개선해보세요.

ghis22130 replied:
JK 질문이 있습니다. "메소드를 일반화 시킨다"라는 말이 잘 이해가 가지 않습니다.
Dealer와 Player에서 모두 똑같이 동작하는 메소드를 일반화 시킨다 라는 것은 상속 관계를 이용한다라고 이해해도 되는걸까요?
아니면 메소드 이름을 똑같이 해주는것(똑같은 기능으로 동작하기 때문에) 일까요?
개념은 알겠지만 구현 방식이 감이 오질 않네요..ㅜ

godrm replied:
동일한 메소드 시그니처를 갖고 있고, 심지어 코드까지 동일하다면 상속 관계를 고민해볼 수 있겠죠.
수퍼클래스에 공통 메소드를 구현하면 그래도 호출만 하면 되니까요.

godrm commented on PokerGameApp/PokerGameApp/PockerGame.swift:
```diff
@@ -8,43 +8,39 @@
+    private var cardStud: Int
     
-    init(numberOfPlayer: Int, cardStud: Int) {
+    init(ofPlayer: Int, cardStud: Int) {
```
Int 타입 대신에 ofPlayer나 cardStud 에 대한 타입을 선언해보세요.
Card.Rank나 Suit 처럼 의미있는 값을 다루는 세부적인 타입을 선언하고 표현하는 게 중요한 부분입니다.

ghis22130 replied:
cardStud는 Enum type 으로 선언 해볼 구상이 생각이 들지만
ofPlayer같은 경우는 전혀 감이 오질 않네요 player의 숫자를 알려주는 type도  Int로 두는건 아니군요.. 단순히 typealias를 이용하는것은 지양해야 하나요?

godrm replied:
네. Int 는 값을 표현할 수 있는 범위가 우리가 다루는 값보다 더 큰 일반적인 타입이죠.
세부적인 타입은 우리가 다루는 값만 표현할 수 있도록 제한하는 겁니다. Card.Rank 나 Crad.Suit 처럼요.

ghis22130 commented:
리뷰 감사합니다. 피드백 반영 후 다시 push 하겠습니다!

ghis22130 committed:
한줄 코딩 수정

ghis22130 committed:
description 위치조정

ghis22130 committed:
변수명 등 전체적인 소소한 리팩토링

ghis22130 committed:
Int type을 대신할 Event, Capacity Type 선언

ghis22130 committed:
Type들을 관리하기 위한 Types.swift 작성

ghis22130 committed:
메소드 일반화를 위한 상속관계 지정

ghis22130 commented:
- Int Type 을 대신 하기 위해 플레이어 인원 -> Capacity로 게임의 인원에 제한
- Event로 게임의 종류 (five, seven) 제한
- Types들을 모아서 관리하기 위해 Types.swift 작성
- 
- 상속관계를 지정해서 메소드 일반화를 시도해보았지만 생각보다 아직은(?) 동일하게 작동하는 부분은 print 부분밖에 발견하지 못했습니다..

godrm commented on PokerGameApp/PokerGameApp/Player.swift:
```diff
@@ -15,11 +15,15 @@ import Foundation
     
-    func receiveCards(with rCards: [Card]) {
+    func receive(with rCards: [Card]) {
         self.cards = rCards
```
rCards는 어떤 의도인지 알기 어렵네요. 이런 문법은 receive(with cards:) 정도로 표현해도 좋겠습니다.

godrm commented on PokerGameApp/PokerGameApp/PockerGame.swift:
```diff
@@ -10,16 +10,14 @@ import Foundation
+    private var cardStud: Event
     
-    init(ofPlayer: Int, cardStud: Int) {
+    init(ofPlayer: Capacity, cardStud: Event) {
```
Capacity는 어느정도 감이 오는데, Event는 타입만 보면 무슨 정보를 표현하는지 모르겠네요.

godrm commented on PokerGameApp/PokerGameApp/Player.swift:
```diff
@@ -15,11 +15,15 @@ import Foundation
+    
+    func printDeck() {
+        print(cards)
+    }
```
player도 마찬가지로 내부 데이터를 직접 print로 출력할 일은 앱을 만드는 동안 없습니다.
앱 화면에 표시하기 위해서는 값을 전달하거나, 접근할 수 있는 메소드를 제공해주는 게 중요합니다. 
내부에서 직접 출력하는 메소드는 앱에서는 불필요하다고 생각하세요.

godrm commented on PokerGameApp/PokerGameApp/Players.swift:
```diff
@@ -20,10 +20,10 @@ struct Players {
             guard let newCards = dealer.give() else { return }
-            player.receiveCards(with: newCards)
+            player.receive(with: newCards)
         }
```
이 부분은 조금 아쉬운데요 Players가 dealer를 모르고 있고 딜러가 (give이후) 넘겨주는 newCard만 채워넣도록 해도 좋았을 것 같습니다.

godrm commented:
수고하셨습니다. 마지막 피드백들은 다음 단계를 하면서 같이 개선해보세요.
